packages/marshal/src/encodePassable.js

```diff
@@ -216,11 +216,131 @@ const decodeBigInt = encoded => {
   return n;
 };
 
-// `'\u0000'` is the terminator after elements.
-// `'\u0001'` is the backslash-like escape character, for
-// escaping both of these characters.
+// Escape all characters from U+0000 NULL to U+001F INFORMATION SEPARATOR ONE
+// like `!<character offset by 0x21>` to avoid JSON.stringify expansion to
+// `\uHHHH`, and specially escape U+0020 SPACE (the array element terminator)
+// as `!_` and U+0021 EXCLAMATION MARK (the escape prefix) as `!|`.
+// Relative lexicographic ordering is preserved by this mapping of any character
+// at or before `!` in the contiguous range [0x00..0x21] to a respective
+// character in [0x21..0x40, 0x5F, 0x7C] preceded by `!` (which is itself in the
+// replaced range).
+// Similarly, escape `^` as `_@` and `_` as `__` because `^` indicates the start
+// of an encoded array.
+const stringEscapes = Array(0x22)
+  .fill(undefined)
+  .map((_, cp) => {
+    switch (String.fromCharCode(cp)) {
+      case ' ':
+        return '!_';
+      case '!':
+        return '!|';
+      default:
+        return `!${String.fromCharCode(cp + 0x21)}`;
+    }
+  });
+stringEscapes['^'.charCodeAt(0)] = '_@';
+stringEscapes['_'.charCodeAt(0)] = '__';
+
+const encodeStringWithEscapes = str =>
+  `s${str.replace(/[\0-!^_]/g, ch => stringEscapes[ch.charCodeAt(0)])}`;
+const decodeStringWithEscapes = encoded => {
+  return encoded.slice(1).replace(/([!_])(.|\n)?/g, (esc, prefix, suffix) => {
+    switch (esc) {
+      case '!_':
+        return ' ';
+      case '!|':
+        return '!';
+      case '_@':
+        return '^';
+      case '__':
+        return '_';
+      default: {
+        const ch = /** @type {string} */ (suffix);
+        (prefix === '!' && ch >= '!' && ch <= '@') ||
+          Fail`invalid string escape: ${q(esc)}`;
+        return String.fromCharCode(ch.charCodeAt(0) - 0x21);
+      }
+    }
+  });
+};
+
+const encodeStringWithoutEscapes = str => `s${str}`;
+const decodeStringWithoutEscapes = encoded => encoded.slice(1);
+
+/**
+ * Encodes an array into a sequence of encoded elements, each terminated by a
+ * space (which is part of the escaped range in encoded strings).
+ *
+ * @param {unknown[]} array
+ * @param {(p: Passable) => string} encodePassable
+ * @returns {string}
+ */
+const encodeArrayWithoutEscapes = (array, encodePassable) => {
+  const chars = ['^'];
+  for (const element of array) {
+    const enc = encodePassable(element);
+    chars.push(enc, ' ');
+  }
+  return chars.join('');
+};
+
+/**
+ * @param {string} encoded
+ * @param {(encoded: string) => Passable} decodePassable
+ * @returns {Array}
+ */
+const decodeArrayWithoutEscapes = (encoded, decodePassable) => {
+  encoded.startsWith('^') || Fail`Encoded array expected: ${encoded}`;
+  const tail = encoded.slice(1);
+  const elements = [];
+  let depth = 0;
+  let nextIndex = 0;
+  let currentElementStart = 0;
+  for (const { 0: ch, index: i } of tail.matchAll(/[\^ ]/g)) {
+    const index = /** @type {number} */ (i);
```

Comment:
Just idly curious why `RegExpMatchArray` type declares its `.index` as optional. Your following unchecked cast relies, seeming correctly, on the knowledge that any match will indeed produce this property. Which would seem to make the optionality unnecessary.

Reply:
I agree, but `lint` doesn't:
```sh
marshal$ yarn lint
yarn run v1.22.19
$ yarn lint:types && yarn lint:eslint
$ tsc -p jsconfig.json
src/encodePassable.js:330:31 - error TS18048: 'index' is possibly 'undefined'.

330         currentElementStart = index + 1;
                                  ~~~~~

src/encodePassable.js:334:17 - error TS18048: 'index' is possibly 'undefined'.

334     nextIndex = index + 1;
```

Even the latest version of TypeScript seems to complain, subject to  `strictNullChecks`: [TS playground](https://www.typescriptlang.org/play?strictNullChecks=true&target=99&ts=5.2.2&filetype=js&useJavaScript=truee=4#code/PQKhAIAEBcE8AcCm4DeBnaAnAlgOwOYC+4IwAUAMYD2uG4iu1AJok+ALzgBEAZlVVwDcZMth7gAFAEIGzVgDoMAQ0zQ0AdWzQAFhIDkAPT0BKY6nA7MVAO7gABgFFGVFmxWYlsegA8kFaKwAXOAAJCiyLqyEdoLghJQ0dNBK2AA2HPTOroqp2BSIEgCMxsLUtND0qYgAtgxqGQDaALrCVRUs8DoZAAytiBW4iN7QAJK4LN49fRUUAK6YmHUOVbW40ADKyapTZHyYkmV0KODdwRTaADTgeBNx4FTiyWny1UrQ5wCCqakSwFTA+FMqDI4Gu4yG4AA1OBCsJCEA)

Reply:
In that case, no change suggested. Resolving.